### ISSUE TITLE
typing docs are unclear on when `AsyncIterator` should be used in type hints as opposed to `AsyncGenerator`

### ISSUE THREAD

**Issue by H0R5E** [python/cpython#112866]
# Bug report

### Bug description:

Hi,

I have an issue with type checking for an aynchronous iterator that I break from for testing purposes. I call the `aclose` method on the iterator to clean it up after the break. This works fine in practice, but typing complains that the method doesn't exist. See the example below (the comment shows the pylance output).

```python
import asyncio
from typing import AsyncIterator


async def ticker(delay: int, to: int) -> AsyncIterator[int]:
    for i in range(to):
        yield i
        await asyncio.sleep(delay)


async def use_ticker():
    
    aiter = ticker(1, 5)
    i = 0
    
    try:
        async for j in aiter:
            if i == 2: break
            print(j)
            i += 1
    finally:
        await aiter.aclose() # Cannot access member "aclose" for type "AsyncIterator[int]"
```

In `collections.abc` the `aclose` method is defined on an `AsyncGenerator`, but `ticker`, above, does not fit that model. `aclose` works on an `AsyncIterator` so I think it should be defined as a method in `collections.abc` also.

Thanks for your help,

Mat

### CPython versions tested on:

3.9

### Operating systems tested on:

Windows

**Comment by H0R5E** [comment]
> What makes you think that `AsyncIterator`s all have `aclose` methods?

Well, you know, I am just following [the docs](https://docs.python.org/3/library/typing.html#typing.AsyncGenerator):

> If your generator will only yield values, set the SendType to None:
> 
> ...
> 
> Alternatively, annotate your generator as having a return type of either AsyncIterable[YieldType] or AsyncIterator[YieldType]

Thanks for your help, anyway.

**Comment by AlexWaygood** [comment]
Hmm. Well, perhaps we could improve the docs there! What the docs are _trying_ to say is "if you're not doing anything with the async generator that's specific to it being an async generator, you can just use the simpler type hint". In this case you _are_ doing something that only async generators have, as opposed to other async iterators -- you're using the `aclose` method.

**Comment by Eclips4** [comment]
@H0R5E Would you like to send a PR with doc update? ;)

**Comment by H0R5E** [comment]
@Eclips4, my proposal for the docs update would be to delete everything from "Alternatively" onwards. It seems to me that just typing these as `AsyncGenerator`s would avoid confusion, given that the generator methods become available when you use the `yield` syntax.

If that's OK, would I submit a PR to all the relevant branches for each Python version?

**Comment by H0R5E** [comment]
@JelleZijlstra, when you say "change your implementation to a different kind of iterable later," do you mean not use `yield` any more? If so, the documentation, as it stands, only gives examples using `yield` in an async function. Can you refactor it without it becoming a class? When you use `yield`, if it's going to become a magic `AsyncGenerator`, why encourage people to use a super type?

@Eclips4, I think the answer is 'no' then, as:

1. I clearly don't know enough to describe the proper use of this case.
2. This alias is deprecated anyway, so I'm not sure that it's worth updating
3. The linked documentation in collections.abc is a one-liner

Thanks again for your help.

Mat

**Comment by graingert** [comment]
@JelleZijlstra I don't think this applies to async generators because at runtime it's very important that the generator be `aclose()`ed deterministically otherwise you get a ResourceWarning when running on trio. See also https://docs.python.org/3/library/contextlib.html#contextlib.aclosing
